inc/class-job.php

```diff
@@ -219,6 +238,166 @@ public static function get_by_site( $site, $include_completed = false, $include_
+			$timestamp = time();
+			$timestamp_compare = '>';
+		} elseif ( $args['timestamp'] === 'past' ) {
+			$timestamp = time();
+			$timestamp_compare = '<=';
+		}
+
+		$args['limit' ] = absint( $args['limit' ] );
+
+		if ( $args['limit'] > 100 ) {
+			trigger_error( 'Exceeding recommended job search limit of 100' );
+		}
+
+		// Find all scheduled events for this site
+		$table = static::get_table();
+
+		$sql = "SELECT * FROM `{$table}` WHERE site = %d";
+		$sql_params[] = $args['site'];
+
+		if ( is_string( $args['hook'] ) ) {
+			$sql .= ' AND hook = %s';
+			$sql_params[] = $args['hook'];
+		}
+
+		if ( ! is_null( $args['args'] ) ) {
+			$sql .= ' AND args = %s';
+			$sql_params[] = serialize( $args['args'] );
+		}
+
+		if (
+			! empty( $args['timestamp'] ) &&
+			in_array( $args['timestamp'], ['past', 'future'] ) &&
+			! empty( $timestamp ) &&
+			! empty( $timestamp_compare ) &&
+			in_array( $timestamp_compare, [ '<=', '<', '>', '>=', '=' ] )
+		) {
+			$sql .= " AND nextrun {$timestamp_compare} %s";
+			$sql_params[] = date( DATE_FORMAT, $timestamp );
+		} elseif (
+			! empty( $args['timestamp'] ) &&
+			is_array( $args['timestamp'] ) &&
+			count( $args['timestamp'] ) === 2
+		) {
+			// It's a range.
+			$sql .= ' AND nextrun BETWEEN %s AND %s';
+			$sql_params[] = date( DATE_FORMAT, $args['timestamp'][0] );
+			$sql_params[] = date( DATE_FORMAT, $args['timestamp'][1] );
+		} elseif ( ! empty( $args['timestamp'] ) ) {
+			$sql .= ' AND nextrun = %s';
+			$sql_params[] = date( DATE_FORMAT, $args['timestamp'] );
+		}
+
+		$sql .= ' AND status IN(' . implode( ',', array_fill( 0, count( $args['statuses'] ), '%s' ) ) . ')';
+		$sql_params = array_merge( $sql_params, $args['statuses'] );
+
+		$sql .= ' ORDER BY nextrun';
+		if ( $args['order'] === 'DESC' ) {
+			$sql .= ' DESC';
+		} else {
+			$sql .= ' ASC';
+		}
+
+		if ( $args['limit'] > 0 ) {
+			$sql .= ' LIMIT %d';
+			$sql_params[] = $args['limit'];
+		}
+
+		$query = $wpdb->prepare( $sql, $sql_params );
+
+		// Use caching if group deletion is supported.
+		if ( function_exists( 'wp_cache_delete_group' ) ) {
+			$query_hash = md5( $query );
+			$found = null;
+			$results = wp_cache_get( "jobs::{$query_hash}", 'cavalcade-job-queries', true, $found );
+			if ( ! $found ) {
+				$results = $wpdb->get_results( $query );
+				wp_cache_set( "jobs::{$query_hash}", $results, 'cavalcade-job-queries' );
+			}
+		} else {
+			$results = $wpdb->get_results( $query );
+		}
+
+		if ( $args['__raw'] === true ) {
+			return $results;
+		}
+		return static::to_instances( $results );
+	}
+
+	/**
+	 * Flush the job query cache. If using WP Redis or persistent store
+	 * that support deleting cache entries by group this will flush the group.
+	 */
+	public static function flush_query_cache() {
+		if ( function_exists( 'wp_cache_delete_group' ) ) {
+			wp_cache_delete_group( 'cavalcade-job-queries' );
+		} else {
+			wp_cache_delete( 'jobs', 'cavalcade-jobs' );
```

Comment:
This only works in tandem with the suggested changes for the query hash.

```suggestion
			wp_cache_delete( 'last_changed', 'cavalcade-job-queries' );
```

Reply:
I need to read up and understand what this does really, haven't come across this pattern yet.

Reply:
It's used throughout core in most of the query classes as faux group invalidation. Here are a couple of examples for the terms API:

https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/class-wp-term-query.php#L698-L702

https://github.com/WordPress/wordpress-develop/blob/a3e674d4f9a212faa47185dff6e82bb4f8aaaac0/src/wp-includes/taxonomy.php#L2560

Reply:
Thanks for the references :)